### PR TITLE
Include attribute name in error messages when it's present

### PR DESCRIPTION
**Pull request by dood-**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  | #630

### CODE REVIEW

File: docs/guide/en/configuring-rules-via-php-attributes.md [hunk 1]
```diff
@@ -74,7 +74,7 @@ final class User
 
 > **Note:** [readonly properties] are supported only starting from PHP 8.1.
 
-Error messages may include an `{attribute}` placeholder that is replaced with the name of the property. If you would 
+Error messages may include an `{label}` placeholder that is replaced with the name of the property. If you would
```
Comment (samdark) [inline docs/guide/en/configuring-rules-via-php-attributes.md:77]:
```suggestion
Error messages may include a `{label}` placeholder that is replaced with the name of the property. If you would 
```

File: docs/guide/en/configuring-rules-via-php-attributes.md [hunk 2]
```diff
@@ -91,6 +91,22 @@ final class User
...
 ```
 
+Alternatively, you can specify a label for all class properties:
+
+```php
+use Yiisoft\Validator\Label;
+use Yiisoft\Validator\Rule\Length;
+use Yiisoft\Validator\Rule\Required;
+
+#[Label('This value')]
```
Comment (samdark) [inline docs/guide/en/configuring-rules-via-php-attributes.md:101]: Why? What's the use-case for that?

**Review by samdark — CHANGES_REQUESTED** [review]

File: docs/guide/en/configuring-rules-via-php-attributes.md (hunk 2 above)
Comment (samdark) [reply docs/guide/en/configuring-rules-via-php-attributes.md:101]: Like "**The value** should be 42"?

Comment (dood-) [reply docs/guide/en/configuring-rules-via-php-attributes.md:101]: "The value should be 42" by default. 
If a class or property has #[Label], then the names of attributes are used

https://github.com/yiisoft/validator/blob/1e73321d83d197b0effb16e5d2c32a95d60913e4/tests/ValidatorTest.php#L1464-L1509

**Review by samdark — APPROVED** [review]

**Comment by dood-** [comment]
Since I added a new placeholder `{label}`, perhaps it is better to make the placeholder `{attribute}` always be replaced by the name of the property?

**Comment by samdark** [comment]
@vjik, @arogachev what do you think?

**Review by vjik — COMMENTED** [review]
Hm. Is there a case when need attribute in message and at the same time, there is a label?

Seems, it's overhead. I suggest use placeholder `{attribute}` only and return translated label, if it exist, or translated attribute otherwise.

**Comment by samdark** [comment]
No, I can't imagine such a case and I agree with @vjik. @dood- would you please adjust the name to be `{attribute}`? Thanks.

**Comment by dood-** [comment]
@samdark, @vjik If we leave "The value" as a default value, then there is a problem with displaying the attribute names of arrays.

Here is a test for an example:
https://github.com/yiisoft/validator/blob/1e73321d83d197b0effb16e5d2c32a95d60913e4/tests/Rule/OneOfTest.php#L172-L180

We can not specify a Label attribute for the array or implement the LabelsProviderInterface, so the placeholder value will always be "The value"

**Comment by vjik** [comment]
> If we leave "The value" as a default value

There will be no default value.

Will be one attribute `{attribute}`. Value of this placeholder by priority:

1) Translated label (if label exist).
2) Translated attribute.

For case with arrays always will be used translated attribute.

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
> If necessary, I can remove Attribute::TARGET_CLASS from the Label attribute.

Yes, makes sense.

File: tests/Rule/AtLeastTest.php [hunk 3]
```diff
@@ -199,71 +199,71 @@ public function dataValidationFailed(): array
...
-                $class,
-                [new AtLeast(['attr1', 'attr2'], min: 2, message: 'Custom message.')],
-                ['' => ['Custom message.']],
-            ],
-            'custom message with parameters' => [
-                $class,
-                [new AtLeast(['attr1', 'attr2'], min: 2, message: 'Attributes - {attributes}, min - {min}.')],
-                ['' => ['Attributes - "attr1", "attr2", min - 2.']],
-            ],
+            //'incorrect input' => [
```
Comment (vjik) [inline tests/Rule/AtLeastTest.php:202]: Does it need to uncomment?

File: src/ValidationContext.php [hunk 4]
```diff
@@ -248,11 +248,7 @@ public function getAttributeLabel(): ?string
      */
     public function getTranslatedAttribute(): ?string
     {
-        if ($this->attribute === null) {
-            return null;
-        }
-
-        $label = $this->attributeLabel ?? $this->attribute;
+        $label = $this->attributeLabel ?? $this->attribute ?? 'The value';
```
Comment (vjik) [inline src/ValidationContext.php:251]: May be use `Value` without `The`? 
It's will be more consistent with attribute names, for example `age`.

Also, suggest for attributes up first symbold, for exmaple `age` → `Age`.

It would be ideal to be able to use several tokens:
• `{attribute}` — `age` or `value`
• `{Attribute}` — `Age` or `Value`

It allow to create pretty message.

**Review by vjik — CHANGES_REQUESTED** [review]

**Comment by dood-** [comment]
> Also need kill mutants from mutation testing.

Is it possible to disable mutants for a specific php version or just use @infection-ignore-all?

**Comment by samdark** [comment]
I guess @vjik meant fixing tests so there are no mutants, not ignoring these.

File: src/Rule/DateTimeHandler.php
Comment (vjik) [inline src/Rule/DateTimeHandler.php:1]: In `DateTime` rule need replace `The {attribute}` to `{Attribute}` into messages.

File: messages/ru/yii-validator.php
Comment (vjik) [inline messages/ru/yii-validator.php:1]: Need adapt other translations also.

File: tests/Rule/DateTimeTest.php [hunk 5]
```diff
@@ -63,42 +63,42 @@ public function dataValidationFailed(): array
...
                 'datetime',
                 [new DateTime(message: 'Attribute - {attribute}, value - {value}.')],
-                ['' => ['Attribute - , value - datetime.']],
+                ['' => ['Attribute - value, value - datetime.']],
             ],
             'empty string and custom message' => [
                 '',
                 [new DateTime()],
-                ['' => ['The  must be a date.']],
+                ['' => ['The value must be a date.']],
```
Comment (vjik) [inline tests/Rule/DateTimeTest.php:91]: Need to replace `The {attribute}` to `{Attribute}` in `DateTime` error messages

**Review by vjik — CHANGES_REQUESTED** [review]
Overall looks good 👍
There's just a little bit left to finish.

**Review by vjik — APPROVED** [review]

**Comment by arogachev** [comment]
Thanks for you great work! I will review this as well.

File: src/ValidationContext.php [hunk 6]
```diff
@@ -243,16 +243,11 @@ public function getAttributeLabel(): ?string
...
      */
-    public function getTranslatedAttribute(): ?string
+    public function getTranslatedAttribute(): string
     {
-        if ($this->attribute === null) {
-            return null;
-        }
-
-        $label = $this->attributeLabel ?? $this->attribute;
+        $label = $this->attributeLabel ?? $this->attribute ?? 'value';
```
Comment (arogachev) [inline src/ValidationContext.php:250]: Maybe make `value` customizable as well?

Comment (arogachev) [inline src/ValidationContext.php:250]: What if `int` attribute names will be used? Will the result be like "0 is not valid."? Here the attribute name can be confused with actual value. Maybe turn it to countable items then? Like "1th item is not valid". Can be discussed in a separate issue and improved in a separate PR.

File: src/Rule/AbstractCompare.php [hunk 7]
```diff
@@ -262,14 +262,14 @@ public function getIncorrectDataSetTypeMessage(): string
...
         return $this->message ?? match ($this->operator) {
-            '==', => 'Value must be equal to "{targetValueOrAttribute}".',
-            '===' => 'Value must be strictly equal to "{targetValueOrAttribute}".',
-            '!=' => 'Value must not be equal to "{targetValueOrAttribute}".',
-            '!==' => 'Value must not be strictly equal to "{targetValueOrAttribute}".',
-            '>' => 'Value must be greater than "{targetValueOrAttribute}".',
-            '>=' => 'Value must be greater than or equal to "{targetValueOrAttribute}".',
-            '<' => 'Value must be less than "{targetValueOrAttribute}".',
-            '<=' => 'Value must be less than or equal to "{targetValueOrAttribute}".',
+            '==', => '{Attribute} must be equal to "{targetValueOrAttribute}".',
```
Comment (arogachev) [inline src/Rule/AbstractCompare.php:265]: `targetValueOrAttribute` - support capitalization as well?

File: src/Rule/AtLeast.php [hunk 8]
```diff
@@ -62,7 +62,7 @@ final class AtLeast implements DumpedRuleInterface, SkipOnErrorInterface, WhenIn
     public function __construct(
         private array $attributes,
         private int $min = 1,
-        private string $incorrectInputMessage = 'The value must be an array or an object.',
+        private string $incorrectInputMessage = '{Attribute} must be an array or an object.',
         private string $message = 'At least {min, number} {min, plural, one{attribute} other{attributes}} from this ' .
```
Comment (arogachev) [inline src/Rule/AtLeast.php:66]: `attributes` - support capitalization as well?

**Review by arogachev — CHANGES_REQUESTED** [review]

File: src/ValidationContext.php (hunk 6 above)
Comment (vjik) [reply src/ValidationContext.php:250]: In what case may this be required?

Comment (vjik) [reply src/ValidationContext.php:250]: The value cannot be `int`. For `int` keys attributes is `null`.

Comment (arogachev) [reply src/ValidationContext.php:250]: Sorry, didn't get you. Could you demonstrate it with an example?

Comment (arogachev) [reply src/ValidationContext.php:250]: It could be useful to customize text for "value" globally (for every error message). This can be completely another word or the same word with adjective added for example, something like "This value" and similar.

Comment (vjik) [reply src/ValidationContext.php:250]: See here: https://github.com/dood-/validator/blob/enh-attribute-name-in-messages/src/Validator.php#L134-L142

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Thank you!
